### PR TITLE
LPS-121753 Drop IE support from build in backportable way

### PR DESCRIPTION
**Pull request by wincent**
> ℹ️ Believe it or not, **this one is safe to backport to 7.3** because we have implemented branch-specific behavior as described [in this PR](https://github.com/liferay/liferay-frontend-projects/pull/214).

Previously, we used to transpile all of our ".js" files to be ES5 compatible. That's a very old standard (5.1 was finalized in 2011), and goes significantly beyond what we strictly needed to in order to meet the requirements of our compatibility matrix.

In this commit, we not only drop support for IE, but we make our Babel config very strictly match exactly what is in the compatibility matrix. As an example of the gains, take frontend-js-web as an example, we see this change strip off about 1,305 lines of code, and 69,752 characters (pre-gzip):

Before:

    $ cloc build/node/packageRunBuild/resources

    -------------------------------------------------------------------------------
    Language                     files          blank        comment           code
    -------------------------------------------------------------------------------
    JavaScript                     139           1589           2345           8142

    $ wc build/node/packageRunBuild/**/*.js | tail -1
       12005   44520  580077 total

After:

    $ cloc build/node/packageRunBuild/resources

    -------------------------------------------------------------------------------
    Language                     files          blank        comment           code
    -------------------------------------------------------------------------------
    JavaScript                     139           1407           2529           6836

    $ wc build/node/packageRunBuild/**/*.js | tail -1
       10701   39260  510325 total

Here is a concrete example of a file, `side_navigation.es.js`:

- Before: https://gist.github.com/wincent/be6203a1e3032251a8974bcc633d3751
- After: https://gist.github.com/wincent/b2b073003ce91b9256e1b86bbb1b8ac2
- Diff: https://gist.github.com/wincent/408131bf38c070b10574e976930ad0aa

The major version bump in `@liferay/npm-scripts` is because in a very narrow _technical_ sense, it is a "breaking change" (ironically, only on 7.4). This is because it requires us to add one .babelrc.js for the benefit of analytics-client-js where we continue to want IE support to be in place, even though it lives on 7.4. The file will be harmless no-op on 7.3, so nothing will break on it.

[Release notes for `@liferay/npm-scripts` v34.0.0](https://github.com/liferay/liferay-frontend-projects/releases/tag/npm-scripts%2Fv34.0.0)

Why so many commits? These correspond to manual changes that were made in response to testing. We had over 360 places where we were using `const` variables before declaring them. This worked before because Babel turned them all into `var`, which doesn't care about the so-called "Temporal Dead Zone". But now that the browser is seeing all these `const`, the use-before-definition sloppiness turns into a `ReferenceError`. So I turned on the ESLint `no-use-before-define` rule and then went hunting down all the errors, and made one commit per OSGi module just in case anything goes wrong (ie. so that we can do a granular revert of the affected module).

### CODE REVIEW

**Comment by wincent** [comment]
ci:test:sf

**Comment by wincent** [comment]
ci:test:relevant

**Comment by wincent** [comment]
ci:test:bundle

**Comment by jbalsas** [comment]
Please be green!! 😂

File: modules/yarn.lock [hunk 1]
```diff
@@ -10603,11 +10603,6 @@ interpret@^2.0.0:
   resolved "https://registry.yarnpkg.com/interpret/-/interpret-2.0.0.tgz#b783ffac0b8371503e9ab39561df223286aa5433"
   integrity sha512-e0/LknJ8wpMMhTiWcjivB+ESwIuvHnBSlBbmP/pSb8CQJldoj1p2qv7xGZ/+BtbTziYRFSz8OsvdbiX45LtYQA==
 
-intersection-observer@0.11.0:
```
Comment (jbalsas) [inline modules/yarn.lock:10606]: Interesting... do you thing this is related to the updated, or a leftover?

Comment (wincent) [reply modules/yarn.lock:10606]: My assumption is that it's a leftover, but I'll check after I've finished my current `ant all`.

**Comment by wincent** [comment]
> Please be green!! 😂

That's what I always say with CI...

please

**Comment by wincent** [comment]
FWIW, I also updated 7.3.x and confirmed that:

- `build` is correctly producing ES5-compatible artifacts (eg. no arrow functions).
- `checkFormat` is correctly complaining about modern syntax like arrows and `const` etc in JSP files.
- if the `.babelrc.js` file gets backported into `analystics-client-js`, it is harmless and doesn't change the build output.

One final thought: the other place that we build an SDK in liferay-portal is remote-app-client-js, but given that we haven't actually published that to the CDN yet, and when we do, it will be from `master`, I think we can deal with making sure that's IE compatible at the time we release it.

**Comment by wincent** [comment]
I'm seeing some errors in the JS console like `TypeError: Liferay.SearchContainer is not a constructor` (and indeed it isn't; if you log `Liferay.SearchContainer.toString()` you get `constructor(config){var id=config.id;config.boundingBox=config.boundingBox||"#"+id;config.contentBox=config.contentBox||"#"+config.id+"SearchContainer";SearchContainer.superclass.constructor.apply(this,arguments)}`) so it's possible that AUI modules don't like the change and we'll need to do some special casing for them (unless of course this is an error in common with `master`, but somehow I doubt that will be the case). I'll dig into that tomorrow.

- `search_container.js` in this PR: https://gist.github.com/wincent/29b623d20bd466c2ba4e545ff58863e3
- `search_container.js` on `master`: https://gist.github.com/wincent/cfc1d89885fd312df0128a99656d278a
- diff: https://gist.github.com/wincent/1ea7e78ce8fc499638369a278d9f0631

Probably relevant, this bit:

```diff
diff --git a/portal-master/liferay-portal/modules/apps/frontend-js/frontend-js-aui-web/build/node/packageRunBuild/resources/liferay/search_container.js b/portal/liferay-portal/modules/apps/frontend-js/frontend-js-aui-web/build/node/packageRunBuild/resources/liferay/search_container.js
index 9c1b05d..669dee4 100644
--- a/portal-master/liferay-portal/modules/apps/frontend-js/frontend-js-aui-web/build/node/packageRunBuild/resources/liferay/search_container.js
+++ b/portal/liferay-portal/modules/apps/frontend-js/frontend-js-aui-web/build/node/packageRunBuild/resources/liferay/search_container.js
@@ -11,7 +11,7 @@
  * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
  * details.
  */
-AUI.add('liferay-search-container', function (A) {
+AUI.add('liferay-search-container', A => {
   var Lang = A.Lang;
   var CSS_TEMPLATE = 'lfr-template';
   var STR_BLANK = '';
@@ -24,13 +24,15 @@ AUI.add('liferay-search-container', function (A) {
       }
     },
     NAME: 'searchcontainer',
-    constructor: function constructor(config) {
+
+    constructor(config) {
       var id = config.id;
       config.boundingBox = config.boundingBox || '#' + id;
       config.contentBox = config.contentBox || '#' + config.id + 'SearchContainer';
       SearchContainer.superclass.constructor.apply(this, arguments);
     },
-    get: function get(id) {
+
+    get(id) {
       var instance = this;
       var searchContainer = null;
 ```

**Comment by wincent** [comment]
I haven't looked deeply into the AUI problem yet, other than to confirm that going back to the old Babel settings in `frontend-js-aui-web` fixes that specific problem with `Liferay.SearchContainer` (I'll analyze this properly later, hopefully without reaching the conclusion that we need to apply a bunch of Babel overrides for the benefit of AUI).

But in the meantime, I have some bad news, as explained in https://github.com/liferay-frontend/liferay-portal/pull/441/commits/c7c44af65ebf5cc13e9ebc07ec31cdd714a4960d 😬 

> Now that we aren't transforming the code for IE, these `const` actually remain as `const` instead of getting turned into `var`, which exposes a use-before-definition problem at runtime:
>
>     ReferenceError: Cannot access 'ExperimentStatusType' before initialization
>
> As a bonus, moving the declaration of `ExperimentStatusType` before its usage ends up putting the exports into alphabetical order.
>
> For reference:
>
> - [`types.js` before dropping IE support](https://gist.github.com/wincent/d5889bfca7242d1093647b824345a785)
> - [`types.js` after dropping IE support](https://gist.github.com/wincent/4ecfb9caa974089fdfd0b9a0eb6ff8e7)
> - [diff](https://gist.github.com/wincent/ca06919af8471c3faeff5120854448d6)
>
> It's possible that we'll run into other subtle issues of this type at runtime; I'm going to see whether [ESLint's `no-use-before-define` rule](https://eslint.org/docs/rules/no-use-before-define) can help us out here, at least for JS in ".js" files (it won't help us in JSPs because we can't treat those as self-contained lexical units).
>
> In a quick test of activating that rule, it finds [over 360 problems](https://gist.github.com/wincent/34bb65530fc4baccfc33922c83e3a79d) that we're going to have to deal with.

TL;DR: this is going to take a while, might have some merge conflicts, and there is some risk involved (because we might break things when we reorder the files); having said that, I think we should definitely proceed, otherwise we'll just drag the technical debt with us forever.

**Comment by wincent** [comment]
Ha, [this one](https://github.com/liferay-frontend/liferay-portal/pull/441/commits/c43ae4bd6c1f8736492b9fa67519559e20d97833) is interesting, at least for me...

> So I spent a bunch of time digging into why AUI doesn't like the object concise method, looking at AUI source code, snooping with a debugger etc, because for all intents and purposes, I couldn't see any real difference between a method defined on an object using "longhand":
>
> ```js
> {
>     a: function a() {},
> }
> ```
>
> and one using concise method syntax:
>
> ```js
> {
>     a() {}
> }
> ```
>
> They have the same name, the same calling behavior, the same enumerability etc. I was suspicious then, of AUI doing something deeply weird deep down inside.
>
> But it turns out that there actually _is_ a difference in calling behavior, which ends up getting manifested in this case as:
>
> ```
> TypeError: Liferay.SearchContainer is not a constructor
> ```
>
> I couldn't find an explanation of this on MDN (it is probably in the spec somewhere though), but I did find this [simple example on Stack Overflow](https://stackoverflow.com/a/45119651/2103996) showing that this is expected behavior for concise methods.
>
> With the benefit of hindsight, we can see that I could have figured this out in five minutes instead of two hours. :facepalm:
>
> Amazingly, there is exactly _one_ place in all of liferay-portal where we actually pass a `constructor` into `A.Component.create`, so there is exactly one place to fix. Found them by searching for `constructor` in all files that contain the string `Component.create`; then did a broader search for `constructor` in all ".js" and ".jsp" files that contain the string `create`. It's possible something a bit obfuscated may have slipped through, but I think this is a reasonable first pass.

**Comment by wincent** [comment]
Ok, I wasn't planning on getting this done today, but here it is - let's see if it's green. I'm going to start the tests running and then mention some folks whose code I've touched.

**Comment by wincent** [comment]
ci:test:sf

**Comment by wincent** [comment]
ci:test:relevant

**Comment by wincent** [comment]
ci:test:bundle

**Comment by wincent** [comment]
Heads up, seeing this involves a lot of manual changes, and I want you all to know who to hunt for with pitchforks and torches if I break anything...

- @ambrinchaudhary, @boton for `ratings-taglib` changes and possibly a few others (see the commit list; is `document-library` yours?).
- @p2kmgcl, @victorg1991, @sandrodw3, @veroglez for changes in `layout-content-page-editor-web`.
- @julien for changes in `frontend-js-aui-web` etc 😈.
- @carloslancha  for changes in `frontend-editor-ckeditor-web` etc 😈.
- @jbalsas for changes in `frontend-image-editor-web`.
- @matuzalemsteles for changes in `dynamic-data-mapping`.

Anybody, please feel free to mention anybody that you know would be relevant for the other modules that I've touched.

**Comment by wincent** [comment]
Seeing as the pull request tester seems to be ignoring commands on this one, I sent an identical copy here for testing: https://github.com/liferay-frontend/liferay-portal/pull/447

**Comment by wincent** [comment]
_Aaaaaaand_ as soon as I said that, the tester seems to have updated... whatever!

**Comment by wincent** [comment]
Ok, seeing as the bot never posted results here I am going to close this one and continue on over at https://github.com/liferay-frontend/liferay-portal/pull/447
